VirtoCommerce.PricingModule.Data/Services/PricingServiceImpl.cs

```diff
@@ -153,34 +152,46 @@ public virtual IEnumerable<coreModel.Price> EvaluateProductPrices(coreModel.Pric
                 prices = query.ToArray().Select(x => x.ToModel(AbstractTypeFactory<coreModel.Price>.TryCreateInstance())).ToArray();
             }
 
-            foreach(var productId in evalContext.ProductIds)
+            var priceListOrdererList = evalContext.PricelistIds?.ToList();
+
+            foreach (var productId in evalContext.ProductIds)
             {
-                var productPrices = prices.Where(x => x.ProductId == productId);              
+                var productPrices = prices.Where(x => x.ProductId == productId);
                 if (evalContext.ReturnAllMatchedPrices)
                 {
                     // Get all prices, ordered by currency and amount.
                     var orderedPrices = productPrices.OrderBy(x => x.Currency).ThenBy(x => Math.Min(x.Sale ?? x.List, x.List));
                     retVal.AddRange(orderedPrices);
                 }
-                else
+                else if (!priceListOrdererList.IsNullOrEmpty())
                 {
-                    //Order by priority
-                    foreach (var currencyPricesGroup in productPrices.GroupBy(x => x.Currency))
+                    // as priceListOrdererList is sorted by priority (descending), we save PricelistId's index as Priority
+                    var priceTuples = productPrices
+                        .Select(x => new { Price = x, x.Currency, x.MinQuantity, Priority = priceListOrdererList.IndexOf(x.PricelistId) })
+                        .Where(x => x.Priority > -1);
+
+                    // Group by Currency and by MinQuantity
+                    foreach (var pricesGroupByCurrency in priceTuples.GroupBy(x => x.Currency))
                     {
-                        var groupPrices = currencyPricesGroup.OrderBy(x => Math.Min(x.Sale ?? x.List, x.List));
-                        if (!evalContext.PricelistIds.IsNullOrEmpty())
+                        var minAcceptablePriority = int.MaxValue;
+                        // take prices with lower MinQuantity first
+                        foreach (var pricesGroupByMinQuantity in pricesGroupByCurrency.GroupBy(x => x.MinQuantity).OrderBy(x => x.Key))
                         {
-                            //return only prices from one prioritized price list
-                            var prioritedPriceListId = evalContext.PricelistIds.FirstOrDefault(x => groupPrices.Any(y => y.PricelistId == x));
-                            if (prioritedPriceListId != null)
+                            // take minimal price from most prioritized Pricelist
+                            var groupAcceptablePrice = pricesGroupByMinQuantity.OrderBy(x => x.Priority)
+                                                                            .ThenBy(x => Math.Min(x.Price.Sale ?? x.Price.List, x.Price.List))
+                                                                            .First();
+
+                            if (minAcceptablePriority >= groupAcceptablePrice.Priority)
```

Comment:
Could you explain the purpose of using this code, because the data is already sorted by priority?

Reply:
data is sorted by MinQuantity and here only prices with higher (or equal) priority are taken